src/audio/crossover/crossover.c

```diff
@@ -0,0 +1,836 @@
+
+static const struct comp_driver comp_crossover;
+
+/* 948c9ad1-806a-4131-ad6c-b2bda9e35a9f */
+DECLARE_SOF_UUID("crossover", crossover_uuid, 0x948c9ad1, 0x806a, 0x4131,
+		 0xad, 0x6c, 0xb2, 0xbd, 0xa9, 0xe3, 0x5a, 0x9f);
+
+static void crossover_free_config(struct sof_crossover_config **config)
+{
+	rfree(*config);
+	*config = NULL;
+}
+
+/**
+ * \brief Reset the state of an LR4 filter.
+ */
+static void crossover_reset_state_lr4(struct iir_state_df2t *lr4)
+{
+	rfree(lr4->delay);
+
+	lr4->coef = NULL;
+	lr4->delay = NULL;
+}
+
+/**
+ * \brief Reset the state (coefficients and delay) of the crossover filter
+ *	  of a single channel.
+ */
+static void crossover_reset_state_ch(struct crossover_state *ch_state)
+{
+	int i;
+
+	for (i = 0; i < CROSSOVER_MAX_LR4; i++) {
+		crossover_reset_state_lr4(&ch_state->lowpass[i]);
+		crossover_reset_state_lr4(&ch_state->highpass[i]);
+	}
+}
+
+/**
+ * \brief Reset the state (coefficients and delay) of the crossover filter
+ *	  across all channels
+ */
+static void crossover_reset_state(struct comp_data *cd)
+{
+	int i;
+
+	for (i = 0; i < PLATFORM_MAX_CHANNELS; i++)
+		crossover_reset_state_ch(&cd->state[i]);
+}
+
+/**
+ * \brief Returns the index i such that assign_sink[i] = pipe_id.
+ *
+ * The assign_sink array in the configuration maps to the pipeline ids.
+ *
+ * \return the position at which pipe_id is found in config->assign_sink.
+ *	   -1 if not found.
+ */
+static uint8_t crossover_get_stream_index(struct sof_crossover_config *config,
+					  uint32_t pipe_id)
+{
+	int i;
+	uint32_t *assign_sink = config->assign_sink;
+
+	for (i = 0; i < config->num_sinks; i++)
+		if (assign_sink[i] == pipe_id)
+			return i;
+
+	comp_cl_err(&comp_crossover, "crossover_get_stream_index() error: couldn't find any assignment for connected pipeline %u",
+		    pipe_id);
+
+	return -1;
+}
+
+/*
+ * \brief Aligns the sinks with their respective assignments
+ *	  in the configuration.
+ *
+ * Refer to sof/src/include/sof/crossover.h for more information on assigning
+ * sinks to an output.
+ *
+ * \param[out] sinks array where the sinks are assigned
+ * \return number of sinks assigned. This number should be equal to
+ *	   config->num_sinks if no errors were found.
+ */
+static int crossover_assign_sinks(struct comp_dev *dev,
+				  struct sof_crossover_config *config,
+				  struct comp_buffer **sinks)
+{
+	struct comp_buffer *sink;
+	struct list_item *sink_list;
+	int num_sinks = 0;
+	int i;
+
+	// align sink streams with their respective configurations
+	list_for_item(sink_list, &dev->bsink_list) {
+		sink = container_of(sink_list, struct comp_buffer, source_list);
+		if (sink->sink->state == dev->state) {
+			// if no config is set, then assign the sinks in order
+			if (!config) {
+				sinks[num_sinks++] = sink;
+				continue;
+			}
+
+			i = crossover_get_stream_index(config,
+						       sink->pipeline_id);
+			if (i < 0) {
+				comp_warn(dev, "crossover_assign_sinks(), could not assign sink %i",
```

Comment:
What has happened to end up here. Should this be an error? Code comments about what kind of possible configuration mistake there is or something would be useful.